flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTable.java

```diff
@@ -605,7 +615,7 @@ public boolean addTask(Task task) throws SlotNotFoundException, SlotNotActiveExc
 	 * @return The removed task if there is any for the given execution attempt id; otherwise null
 	 */
 	public Task removeTask(ExecutionAttemptID executionAttemptID) {
-		checkInit();
+		checkStartedOrClosing();
```

Comment:
Are we sure that `removeTask` can only be called when the state is started or closing? Can't it happen that we call `removeTask` if there is a late `TaskExecutor#updateTaskExecutionState` call?

Reply:
All `TaskExecutor#updateTaskExecutionState` callbacks with `removeTask` should be scheduled to the main thread in task executor before task termination future is competed. This means that the final `CLOSED` table state will not be set before that. If so, I would keep the check.